src/write_influxdb_udp.c

```diff
@@ -0,0 +1,701 @@
+        continue;
+      if (have_values)
+        BUFFER_ADD(",");
+      BUFFER_ADD("%s=%lf", ds->ds[i].name, vl->values[i].gauge);
+      have_values = true;
+    } else if (wifxudp_config_store_rates) {
+      if (rates == NULL)
+        rates = uc_get_rate(ds, vl);
+      if (rates == NULL) {
+        WARNING("write_influxdb_udp plugin: "
+                "uc_get_rate failed.");
+        return -1;
+      }
+      if (isnan(rates[i]))
+        continue;
+      if (have_values)
+        BUFFER_ADD(",");
+      BUFFER_ADD("%s=%lf", ds->ds[i].name, rates[i]);
+      have_values = true;
+    } else if (ds->ds[i].type == DS_TYPE_COUNTER) {
+      if (have_values)
+        BUFFER_ADD(",");
+      BUFFER_ADD("%s=%" PRIu64 "i", ds->ds[i].name,
+                 (uint64_t)vl->values[i].counter);
+      have_values = true;
+    } else if (ds->ds[i].type == DS_TYPE_DERIVE) {
+      if (have_values)
+        BUFFER_ADD(",");
+      BUFFER_ADD("%s=%" PRIi64 "i", ds->ds[i].name, vl->values[i].derive);
+      have_values = true;
+    } else if (ds->ds[i].type == DS_TYPE_ABSOLUTE) {
+      if (have_values)
+        BUFFER_ADD(",");
+      BUFFER_ADD("%s=%" PRIu64 "i", ds->ds[i].name, vl->values[i].absolute);
+      have_values = true;
+    }
+
+  } /* for ds->ds_num */
+  sfree(rates);
+
+  if (!have_values) {
+    return 0;
+  }
+  BUFFER_ADD(" %" PRIu64 "\n", CDTIME_T_TO_MS(vl->time));
+
+#undef BUFFER_ESCAPE
+#undef BUFFER_ADD
+
+  return offset;
+} /* int write_influxdb_point */
+
+static int
+write_influxdb_udp_write(const data_set_t *ds, const value_list_t *vl,
+                         user_data_t __attribute__((unused)) * user_data) {
+  int status;
+  char *values;
+  int values_len;
+
+  /* listen_loop is set to non-zero in the shutdown callback, which is
+   * guaranteed to be called *after* all the write threads have been shut
+   * down. */
+  assert(listen_loop == 0);
+
+  if (wifxudp_config_packet_size < NET_DEFAULT_PACKET_SIZE)
+    values_len = wifxudp_config_packet_size;
+  else
+    values_len = NET_DEFAULT_PACKET_SIZE;
+
+  values = malloc(values_len);
+  if (values == NULL) {
+    ERROR("write_influxdb_udp plugin: malloc failed.");
+    return -1;
+  }
+
+  status = write_influxdb_point(values, values_len, ds, vl);
+  if (status == 0) {
+    /* no real values to send (nan) */
+    free(values);
+    return 0;
+  }
+  if (status < 0) {
+    ERROR("write_influxdb_udp plugin: write_influxdb_point failed.");
+    free(values);
+    return -1;
+  } else if (status > wifxudp_config_packet_size) {
+    ERROR("write_influxdb_udp plugin: too long values.");
+    free(values);
+    return -1;
+  }
+
+  pthread_mutex_lock(&send_buffer_lock);
+
+  if (wifxudp_config_packet_size - send_buffer_fill < status)
+    flush_buffer();
+
+  memcpy(send_buffer_ptr, values, status);
+  send_buffer_fill += status;
+  send_buffer_ptr += status;
+  send_buffer_last_update = cdtime();
+
+  if (wifxudp_config_packet_size - send_buffer_fill < 120)
```

Comment:
What is magic of this `120` value?

Reply:
It's the measured average point size in my testing. As plugin, type, instance, and timestamps have the usual values, the differences with others setups are mostly hostnames, so it should not be very different.

Due the nature of this protocol, it probably should be used throught the loopback interface and with a bigger than usual packet size, so it seemed a good approach to me, but it's OK to remove also.

Reply:
I think we need to add some comment about choosed value and goals.

Reply:
And another question: Why you don't use principle from `network` plugin, it write values to output buffer directly, w/o intermediate buffer + memcpy(), and does retry if data does not fit to buffer.
That avoids `memcpy()` for each written value, in cost of retry.

Reply:
I though that in network plugin the size of the value part was predictable, but in the end you are right, I think it can be done. Please let me some time to rewrite and test it.

Reply:
I've committed a new implementation writing directly to udp buffer.